src/lib.rs

```diff
@@ -251,42 +255,75 @@ impl EndpointHeader {
-                access.write(self.0, offset, 0xffffffff);
-                let mut readback = access.read(self.0, offset);
-                access.write(self.0, offset, address);
-
-                /*
-                 * If the entire readback value is zero, the BAR is not implemented, so we return `None`.
-                 */
-                if readback == 0x0 {
-                    return None;
-                }
-
-                readback.set_bits(0..4, 0);
-                1 << readback.trailing_zeros()
-            };
-
-            match bar.get_bits(1..3) {
-                0b00 => Some(Bar::Memory32 {
-                    address,
-                    size,
-                    prefetchable,
-                }),
+            match bar.get_bits(1..3)
+            {
+                0b00 => {
+                    let size = unsafe {
+                        access.write(self.0, offset, 0xfffffff0);
+                        let mut readback = access.read(self.0, offset).get_bits(4..32);
+                        access.write(self.0, offset, address);
+
+                        if readback == 0x0 {
+                            return None;
+                        }
+
+                        readback.set_bits(0..4, 0);
+                        1 << readback.trailing_zeros()
+                    };
+                    Some(Bar::Memory32 {
+                        address,
+                        size,
+                        prefetchable,
+                    })
+                },
+
                 0b10 => {
+                    /*
+                     * If the BAR is 64 bit-wide and this slot is the last, there is no second slot to read.
+                     */
+                    if slot >= 5 {
+                        return None
+                    }
+
+                    let address_upper = unsafe { access.read(self.0, offset + 4) };
+
+                    let size = unsafe {
+                        access.write(self.0, offset, 0xfffffff0);
+                        access.write(self.0, offset + 4, 0xffffffff);
+                        let readback_low = access.read(self.0, offset);
+                        let readback_high = access.read(self.0, offset + 4);
+                        access.write(self.0, offset, address);
+                        access.write(self.0, offset + 4, address_upper);
+
+                        /*
+                         * If the readback from the first slot is not 0, the size of the BAR is less than 4GiB.
+                         */
+                        if readback_low != 0
+                        {
+                            /*
+                             * The readback is invalid in these conditions:
+                             */
+                            (1 << readback_low.trailing_zeros()) as u64
+                        }
+                        else
+                        {
+                            (1 << (readback_high.trailing_zeros() + 32)) as u64
```

Comment:
I think this will overflow, as it'll try to do it on a `u32` by default? You probably need to use `1u64` as the constant and cast the inner bit to `u64`.